malariagen_data/anoph/cnv_data.py

```diff
@@ -0,0 +1,885 @@
+        inline_array: base_params.inline_array = base_params.inline_array_default,
+        chunks: base_params.chunks = base_params.chunks_default,
+    ):
+        debug = self._log.debug
+
+        debug("normalise parameters")
+        sample_sets = self._prep_sample_sets_param(sample_sets=sample_sets)
+        regions: List[Region] = parse_multi_region(self, region)
+        del region
+
+        debug("access CNV HMM data and concatenate as needed")
+        lx = []
+        for r in regions:
+            ly = []
+            for s in sample_sets:
+                y = self._cnv_hmm_dataset(
+                    contig=r.contig,
+                    sample_set=s,
+                    inline_array=inline_array,
+                    chunks=chunks,
+                )
+                ly.append(y)
+
+            debug("concatenate data from multiple sample sets")
+            x = simple_xarray_concat(ly, dim=DIM_SAMPLE)
+
+            debug("handle region, do this only once - optimisation")
+            if r.start is not None or r.end is not None:
+                start = x["variant_position"].values
+                end = x["variant_end"].values
+                index = pd.IntervalIndex.from_arrays(start, end, closed="both")
+                # noinspection PyArgumentList
+                other = pd.Interval(r.start, r.end, closed="both")
+                loc_region = index.overlaps(other)
+                x = x.isel(variants=loc_region)
+
+            lx.append(x)
+
+        debug("concatenate data from multiple regions")
+        ds = simple_xarray_concat(lx, dim=DIM_VARIANT)
+
+        debug("handle sample query")
+        if sample_query is not None:
+            debug("load sample metadata")
+            df_samples = self.sample_metadata(sample_sets=sample_sets)
+
+            debug("align sample metadata with CNV data")
+            cnv_samples = ds["sample_id"].values.tolist()
+            df_samples_cnv = (
+                df_samples.set_index("sample_id").loc[cnv_samples].reset_index()
+            )
+
+            debug("apply the query")
+            loc_query_samples = df_samples_cnv.eval(sample_query).values
+            if np.count_nonzero(loc_query_samples) == 0:
+                raise ValueError(f"No samples found for query {sample_query!r}")
+
+            ds = ds.isel(samples=loc_query_samples)
+
+        debug("handle coverage variance filter")
+        if max_coverage_variance is not None:
+            cov_var = ds["sample_coverage_variance"].values
+            loc_pass_samples = cov_var <= max_coverage_variance
+            ds = ds.isel(samples=loc_pass_samples)
+
+        return ds
+
+    @check_types
+    @doc(
+        summary="Open CNV coverage calls zarr.",
+        returns="Zarr hierarchy.",
+    )
+    def open_cnv_coverage_calls(
+        self, sample_set: base_params.sample_set, analysis: hap_params.analysis
```

Comment:
I feel like we shouldn't be using the `hap_params.analysis` type here, but should define a new type like `cnv_params.coverage_calls_analysis`. This is because the possible values will be slightly different, e.g., there is no "gamb_colu_arab" CNV analysis in Ag3 IIRC.

Reply:
I did wonder about that, thanks. Will do.